nanoFirmwareFlasher.Tool/NanoDeviceManager.cs

```diff
@@ -38,51 +39,55 @@ public async Task<ExitCodes> ProcessAsync()
                 return ExitCodes.E6001;
             }
 
-            NanoDeviceOperations _nanoDeviceOperations = new NanoDeviceOperations();
-
-            if (_options.DeviceDetails)
+            if (!(_options.DeviceDetails || _options.Update || _options.Deploy))
             {
-                NanoDeviceBase nanoDevice = null;
-                return _nanoDeviceOperations.GetDeviceDetails(
-                    _options.SerialPort,
-                    ref nanoDevice);
+                throw new NoOperationPerformedException();
             }
-            else if (_options.Update)
-            {
-                exitCode = await _nanoDeviceOperations.UpdateDeviceClrAsync(
-                    _options.SerialPort,
-                    _options.FwVersion,
-                    _options.ClrFile,
-                    _verbosityLevel);
 
-                if (exitCode != ExitCodes.OK)
-                {
-                    return exitCode;
-                }
-
-                // flag operation as done
-                failedToDoSomething = false;
-            }
+            NanoDeviceOperations _nanoDeviceOperations = new NanoDeviceOperations();
 
-            if (_options.Deploy)
+            if (!GlobalExclusiveDeviceAccess.CommunicateWithDevice(_options.SerialPort, () =>
             {
-                exitCode = _nanoDeviceOperations.DeployApplication(
-                    _options.SerialPort,
-                    _options.DeploymentImage,
-                    _verbosityLevel);
-
-                if (exitCode != ExitCodes.OK)
+                if (_options.DeviceDetails)
                 {
-                    return exitCode;
+                    NanoDeviceBase nanoDevice = null;
+                    exitCode = _nanoDeviceOperations.GetDeviceDetails(
+                        _options.SerialPort,
+                        ref nanoDevice);
+                    return;
                 }
+                else if (_options.Update)
+                {
+                    exitCode = _nanoDeviceOperations.UpdateDeviceClrAsync(

```

Comment:
missing the `await` here

Reply:
The reason is the same as for the PR I submitted after this commit: https://github.com/nanoframework/nf-debugger/pull/377
Ans I think the solution is to make the `() => {}` argument to  GlobalExclusiveDeviceAccess.CommunicateWithDevice into a `async () => {}` and put the await back.

Reply:
Agreed. This one has to be awaitable.
I suppose you can improve this after the PR on debugger is merged. Correct?